### PR TITLE
Cuda kernels for Lower triangular solve

### PR DESCRIPTION
**Pull request by pratikvn**
This PR implements the CUDA kernels for the lower triangular solver. 

For CUDA versions <=9.1, one can only use [csrsm_solve](https://docs.nvidia.com/cuda/archive/9.1/cusparse/index.html#cusparse-lt-t-gt-csrsmsolve)

For CUDA versions >=9.2, one can use [csrsm2_solve](https://docs.nvidia.com/cuda/archive/9.2/cusparse/index.html#cusparse-lt-t-gt-csrsm2_solve), and this algorithm is more efficient and the previous csrsm_solve is [deprecated](https://docs.nvidia.com/cuda/archive/10.1/cusparse/index.html#csrsm_solve) from 10.1, so, unfortunately, we have to do a switch using `#if defines`.

Additionally, CUDA uses different algorithms than the simple algorithms implemented in the reference and omp kernels. Hence I am not sure if it makes sense to compare the kernels as we do for other solvers.

Also, I am not sure how to test the generate kernel properly (or of it is even testable), because all it does it allocate and create from information required for csrsm2_solve calls.

There are additional parameters that can be set for example the `algo` and the `CUSPARSE_SOLVE_POLICY_USE_LEVEL` [(see documentation)](https://docs.nvidia.com/cuda/archive/10.0/cusparse/index.html#cusparse-lt-t-gt-csrsm2_solve), which technically should be exposed to the user, as they may want to tweak it, but I dont do it yet as it can complicate the parameter choices. But I could add these if required.

Note: Currently, only single right hand side version works. I tried but I cannot get the multiple rhs to work. If someone has an idea of how to solve this, I am happy to discuss. The things that I have tried:

1. Because Cusparse uses col-major rather than row major (as Ginkgo does), I tried (as @yhmtsai  suggested) to use `CUSPARSE_OPERATION_NON_TRANSPOSE` for the rhs. But that does not work as well. I think maybe this is not a problem as the single right hand side seems to work for the `CUSPARSE_OPERATION_NON_TRANSPOSE` ans as you actually pass in the both dimensions of the rhs the cusparse function can actually figure out and the right thing.

2. I also tried transposing the rhs and the sol matrix before hand and passing them to the cusparse function with the non-transpose now, but this also has the same problem.

Update: For future references: The multiple RHS as expected now works on CUDA>=9.2, but for CUDA versions <=9.1, multiple rhs solves are handled in a loop with each loop doing 1 rhs solve.

### CODE REVIEW

**Comment by tcojean** [comment]
If you want both better coverage result and to check your code for Intel compilers (should be fine though) you could rebase to the latest develop.

File: cuda/solver/lower_trs_kernels.cu [hunk 1]
```diff
@@ -53,10 +58,132 @@ namespace cuda {
...
+
+
 template <typename ValueType, typename IndexType>
 void generate(std::shared_ptr<const CudaExecutor> exec,
               const matrix::Csr<ValueType, IndexType> *matrix,
-              const matrix::Dense<ValueType> *b) GKO_NOT_IMPLEMENTED;
+              const gko::size_type num_rhs)
+{
+    if (cusparse::is_supported<ValueType, IndexType>::value) {
+        ValueType one = 1.0;
```
Comment (yhmtsai) [inline cuda/solver/lower_trs_kernels.cu]: This variable is unused

File: cuda/solver/lower_trs_kernels.cu [hunk 2]
```diff
@@ -53,10 +58,132 @@ namespace cuda {
...
+            cusparseCreateMatDescr(&cusp_csrsm_data.factor_descr));
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseSetMatIndexBase(
+            cusp_csrsm_data.factor_descr, CUSPARSE_INDEX_BASE_ZERO));
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseSetMatType(
+            cusp_csrsm_data.factor_descr, CUSPARSE_MATRIX_TYPE_GENERAL));
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseSetMatDiagType(
+            cusp_csrsm_data.factor_descr, CUSPARSE_DIAG_TYPE_NON_UNIT));
+#endif
+
+#if (defined(CUDA_VERSION) && (CUDA_VERSION > 9100))
```
Comment (yhmtsai) [inline cuda/solver/lower_trs_kernels.cu]: `CUDA_VERSION` is 9010 for cuda9.1
`CUDA_VERSION` is 9020 for cuda9.2

File: omp/solver/lower_trs_kernels.cpp [hunk 3]
```diff
@@ -58,10 +58,17 @@ namespace omp {
 namespace lower_trs {
 
 
+void clear(std::shared_ptr<const OmpExecutor> exec)
+{
+    // This clear kernel is here to allow for a more sophisticated
```
Comment (thoasm) [inline omp/solver/lower_trs_kernels.cpp]: It is missing what the "more sophisticated implementation" can actually do with this function.
```suggestion
    // This clear kernel is here to allow cleanup for a more sophisticated
```

File: cuda/solver/lower_trs_kernels.cu [hunk 4]
```diff
@@ -53,10 +58,124 @@ namespace cuda {
...
+#if (defined(CUDA_VERSION) && (CUDA_VERSION > 9010))
+struct cusp_csrsm2_data {
+    int algorithm;
+    csrsm2Info_t solve_info;
+    cusparseSolvePolicy_t policy;
+    cusparseMatDescr_t factor_descr;
+    size_t factor_work_size;
+    void *factor_work_vec;
+};
+static cusp_csrsm2_data cusp_csrsm2_data{};
```
Comment (thoasm) [inline cuda/solver/lower_trs_kernels.cu]: Please don't use global variables, especially for data that is tied to an executor.
This prevents us from using multiple CUDA executors at the same time (while doing TRS).

Also: Why are you using the same name for the `struct` and the object?

I see 3 options:
- store the information inside the CUDA executor
- store it inside the LowerTrs class
- (also not a fan of it, but definitively better than the current state:) make a `std::map` with objects of the `*_data` structs tied to the corresponding executor.

Even the last option would be better than now, but I would prefer the first two options (or others that I did not think about).

File: cuda/solver/lower_trs_kernels.cu [hunk 5]
```diff
@@ -53,10 +58,124 @@ namespace cuda {
...
+    cusparseSolveAnalysisInfo_t solve_info;
+    cusparseMatDescr_t factor_descr;
+};
+static cusp_csrsm_data cusp_csrsm_data{};
+#endif
+
+
+void clear(std::shared_ptr<const CudaExecutor> exec)
+{
+#if (defined(CUDA_VERSION) && (CUDA_VERSION > 9010))
```
Comment (thoasm) [inline cuda/solver/lower_trs_kernels.cu]: Why are you doing `CUDA_VERSION < 9020` and `CUDA_VERSION > 9010`? What about the numbers between `9010` and `9020`? I think it would be a lot easier to understand (and be future proof in case CUDA adds these numbers) to do:
`CUDA_VERSION < 9020` and
`CUDA_VERSION >= 9020`.

This applies to all of the `#if`s in this file.

File: cuda/solver/lower_trs_kernels.cu [hunk 6]
```diff
@@ -53,10 +58,124 @@ namespace cuda {
...
+            cusparseCreateMatDescr(&cusp_csrsm_data.factor_descr));
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseSetMatIndexBase(
+            cusp_csrsm_data.factor_descr, CUSPARSE_INDEX_BASE_ZERO));
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseSetMatType(
+            cusp_csrsm_data.factor_descr, CUSPARSE_MATRIX_TYPE_GENERAL));
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseSetMatDiagType(
+            cusp_csrsm_data.factor_descr, CUSPARSE_DIAG_TYPE_NON_UNIT));
+#endif
+
+#if (defined(CUDA_VERSION) && (CUDA_VERSION > 9010))
```
Comment (thoasm) [inline cuda/solver/lower_trs_kernels.cu]: Why are you opening another `#if` instead of writing the code inside the `#if` above with the exact same condition?
You are not doing anything in between both `#if`s.

File: cuda/solver/lower_trs_kernels.cu [hunk 7]
```diff
@@ -53,10 +58,124 @@ namespace cuda {
...
+            matrix->get_const_row_ptrs(), matrix->get_const_col_idxs(), nullptr,
+            num_rhs, cusp_csrsm2_data.solve_info, cusp_csrsm2_data.policy,
+            &cusp_csrsm2_data.factor_work_size);
+
+        // allocate workspace
+        if (cusp_csrsm2_data.factor_work_vec != nullptr) {
+            exec->free(cusp_csrsm2_data.factor_work_vec);
+        }
+        cusp_csrsm2_data.factor_work_vec =
+            exec->alloc<void *>(cusp_csrsm2_data.factor_work_size);
```
Comment (thoasm) [inline cuda/solver/lower_trs_kernels.cu]: Please move these allocations inside the Constructor of the `struct` (and deletion inside the Destructor), otherwise, you can leak memory if somethings throws inside this (or any other) function before calling `free`.

Comment (pratikvn) [reply cuda/solver/lower_trs_kernels.cu]: Unfortunately, I cannot move these inside the constructors as the allocations are dependent on the data from the buffer_size_ext cusparse function and hence can only be allocated after that function has been called.

Comment (thoasm) [reply cuda/solver/lower_trs_kernels.cu]: Then at least move the `free` to the destructor of the `struct`, that way, it should still be cleaned up, even if something throws inside this or another function.

Comment (pratikvn) [reply cuda/solver/lower_trs_kernels.cu]: The destructor of the TRS solver should be called in case some function throws, right ? The `free` is in the call to that destructor, within the `clear` kernel. I checked an atleast here for this struct (maybe I should change the struct somehow) the destructor is not called even if I add one.

Comment (thoasm) [reply cuda/solver/lower_trs_kernels.cu]: That is correct, currently, it should be called in the destructor of the TRS class.
But usually, classes and structs should be responsible for the memory that is allocated inside them, which is why I would prefer to delete the memory in the destructor of the `struct`.
That also prevents programming errors (e.g. if someone wants to extend the current state, and forgets a single `clear` or `free`).

The destructor of the struct as it is now will only (if at all) be called when the whole program terminates since it is as global variable.

File: cuda/test/solver/lower_trs_kernels.cpp [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/solver/lower_trs.hpp>
+
+
+#include <gtest/gtest.h>
```
Comment (yhmtsai) [inline cuda/test/solver/lower_trs_kernels.cpp]: `#include <gtest/gtest.h>` is third party header, so it should be moved to the section between cuda and ginkgo public header in the [guideline](https://github.com/ginkgo-project/ginkgo/wiki/Contributing-guidelines#include-statement-grouping)

File: cuda/test/solver/lower_trs_kernels.cpp [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+    d_b2->copy_from(b.get());
+    b2->copy_from(b.get());
+
+    auto lower_trs_factory = gko::solver::LowerTrs<>::build().on(ref);
+    auto d_lower_trs_factory = gko::solver::LowerTrs<>::build().on(cuda);
+    auto solver = lower_trs_factory->generate(csr_mtx);
+    auto d_solver = d_lower_trs_factory->generate(d_csr_mtx);
+    solver->apply(b2.get(), x.get());
+    d_solver->apply(d_b2.get(), d_x.get());
+    GKO_ASSERT_MTX_NEAR(d_x, x, 1e-14);
```
Comment (yhmtsai) [inline cuda/test/solver/lower_trs_kernels.cpp]: nit: add blank between arrange, act, and assert to follow AAA pattern

File: include/ginkgo/core/base/executor.hpp [hunk 10]
```diff
@@ -883,6 +883,7 @@ class CudaExecutor : public detail::ExecutorBase<CudaExecutor>,
         return cusparse_handle_.get();
     }
 
+
```
Comment (yhmtsai) [inline include/ginkgo/core/base/executor.hpp]: do not need this blank line

File: omp/test/solver/lower_trs_kernels.cpp [hunk 11]
```diff
@@ -105,23 +121,40 @@ class LowerTrs : public ::testing::Test {
...
     std::shared_ptr<Mtx> d_mat;
     std::shared_ptr<CsrMtx> d_csr_mat;
+    std::shared_ptr<gko::solver::SolveStruct> solve_struct;
 };
 
 
+TEST_F(LowerTrs, OmpLowerTrsFlagCheckIsCorrect)
+{
+    bool trans_flag = true;
+    bool expected_flag = false;
```
Comment (yhmtsai) [inline omp/test/solver/lower_trs_kernels.cpp]: need a blank line

File: omp/test/solver/lower_trs_kernels.cpp [hunk 12]
```diff
@@ -132,8 +165,8 @@ TEST_F(LowerTrs, ApplyIsEquivalentToRef)
     initialize_data(59, 3);
```
Comment (yhmtsai) [inline omp/test/solver/lower_trs_kernels.cpp]: need blank line

Comment (pratikvn) [reply omp/test/solver/lower_trs_kernels.cpp]: I think here it is according to the AAA pattern.

Comment (yhmtsai) [reply omp/test/solver/lower_trs_kernels.cpp]: you are right. it is an arrange part.

File: cuda/base/cusparse_bindings.hpp [hunk 13]
```diff
@@ -491,6 +516,317 @@ inline void destroy(cusparseMatDescr_t descr)
...
+        cusparseCreateSolveAnalysisInfo(&solve_struct->solve_info));
+    GKO_ASSERT_NO_CUSPARSE_ERRORS(
+        cusparseCreateMatDescr(&solve_struct->factor_descr));
+    GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseSetMatIndexBase(
+        solve_struct->factor_descr, CUSPARSE_INDEX_BASE_ZERO));
+    GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseSetMatType(
+        solve_struct->factor_descr, CUSPARSE_MATRIX_TYPE_GENERAL));
+    GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseSetMatDiagType(
+        solve_struct->factor_descr, CUSPARSE_DIAG_TYPE_NON_UNIT));
+#endif
```
[the remark this answers is not in the document]
Comment (tcojean) [reply cuda/base/cusparse_bindings.hpp]: Another option is to not put this as constructor but at least as member function. Same for the destructor part.

File: cuda/test/solver/lower_trs_kernels.cpp [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+
+
+#include <memory>
+#include <random>
+
+
+#include <gtest/gtest.h>
+
+
+#include <cuda.h>
```
Comment (yhmtsai) [inline cuda/test/solver/lower_trs_kernels.cpp]: I am not sure whether gtest is at the right location. I created a issue for it #343

File: cuda/solver/lower_trs_kernels.cu [hunk 15]
```diff
@@ -65,8 +152,60 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
...
-           matrix::Dense<ValueType> *x) GKO_NOT_IMPLEMENTED;
+           gko::solver::SolveStruct *solve_struct,
+           matrix::Dense<ValueType> *trans_b, matrix::Dense<ValueType> *trans_x,
+           const matrix::Dense<ValueType> *b, matrix::Dense<ValueType> *x)
+{
+    using vec = matrix::Dense<ValueType>;
+    if (cusparse::is_supported<ValueType, IndexType>::value) {
+        ValueType one = 1.0;
+        auto handle = exec->get_cusparse_handle();
+#if (defined(CUDA_VERSION) && (CUDA_VERSION >= 9020))
```
Comment (tcojean) [inline cuda/solver/lower_trs_kernels.cu]: same thing with the blanklines around the guards

File: cuda/test/solver/lower_trs_kernels.cpp [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+
+
+namespace {
+
+
+class LowerTrs : public ::testing::Test {
+protected:
+    using CsrMtx = gko::matrix::Csr<double, gko::int32>;
+    using Mtx = gko::matrix::Dense<>;
+    LowerTrs() : rand_engine(30) {}
```
Comment (tcojean) [inline cuda/test/solver/lower_trs_kernels.cpp]: blank line above

File: cuda/test/solver/lower_trs_kernels.cpp [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+};
+
+
+TEST_F(LowerTrs, CudaLowerTrsFlagCheckIsCorrect)
+{
+    bool trans_flag = true;
+    bool expected_flag = false;
+#if (defined(CUDA_VERSION) && (CUDA_VERSION >= 9020))
+    expected_flag = false;
+#elif (defined(CUDA_VERSION) && (CUDA_VERSION < 9020))
```
Comment (tcojean) [inline cuda/test/solver/lower_trs_kernels.cpp]: technically since it's already false you only need to test this one

File: cuda/test/solver/lower_trs_kernels.cpp [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+#endif
+    gko::kernels::cuda::lower_trs::perform_transpose(cuda, trans_flag);
+
+    ASSERT_EQ(expected_flag, trans_flag);
+}
+
+
+TEST_F(LowerTrs, CudaSingleRhsApplyIsEquivalentToRef)
+{
+    std::shared_ptr<Mtx> mtx = gen_l_mtx(50, 50);
```
Comment (tcojean) [inline cuda/test/solver/lower_trs_kernels.cpp]: You should move some of these definitions, at least the `mtx` as long as they can be shared into the `LowerTrs` test class.

File: cuda/base/cusparse_bindings.hpp [hunk 19]
```diff
@@ -491,6 +516,317 @@ inline void destroy(cusparseMatDescr_t descr)
...
+            cusparseDestroyCsrsm2Info(solve_struct->solve_info));
+    }
+    if (solve_struct->factor_work_vec != nullptr) {
+        GKO_ASSERT_NO_CUDA_ERRORS(cudaFree(solve_struct->factor_work_vec));
+    }
+#elif (defined(CUDA_VERSION) && (CUDA_VERSION < 9020))
+    cusparse::destroy(solve_struct->factor_descr);
+    GKO_ASSERT_NO_CUSPARSE_ERRORS(
+        cusparseDestroySolveAnalysisInfo(solve_struct->solve_info));
+#endif
```
[the remark this answers is not in the document]
Comment (tcojean) [reply cuda/base/cusparse_bindings.hpp]: They are for example different from the cusparse handle thing in that the custom deallocation function we have is about how to delete the handle itself. Whereas here this is purely internal things, so they should stay internal.

File: cuda/base/cusparse_bindings.hpp [hunk 20]
```diff
@@ -44,6 +45,30 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
+struct SolveStruct {
+    int algorithm;
+    csrsm2Info_t solve_info;
+    cusparseSolvePolicy_t policy;
+    cusparseMatDescr_t factor_descr;
+    size_t factor_work_size;
+    void *factor_work_vec;
+};
+
+#elif (defined(CUDA_VERSION) && (CUDA_VERSION < 9020))
```
[the remark this answers is not in the document]
Comment (pratikvn) [reply cuda/base/cusparse_bindings.hpp]: you mean one blank line above and one blank line below ?

[the remark this answers is not in the document]
Comment (tcojean) [reply cuda/base/cusparse_bindings.hpp]: See for example `core/base/extended_float.hpp`, the top lines.

File: cuda/base/cusparse_bindings.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (pratikvn) [reply cuda/base/cusparse_bindings.hpp]: Thanks, this is a good idea

File: cuda/test/solver/lower_trs_kernels.cpp (hunk 17 above)
Comment (pratikvn) [reply cuda/test/solver/lower_trs_kernels.cpp]: Yes, that is true. But I want to be explicit that it should be false.

File: cuda/test/solver/lower_trs_kernels.cpp (hunk 18 above)
Comment (pratikvn) [reply cuda/test/solver/lower_trs_kernels.cpp]: I would prefer if these would be separate as then you dont have any overlaps between the different tests. For the run-time, I think it would not hurt as the matrices are pretty small.

Comment (tcojean) [reply cuda/test/solver/lower_trs_kernels.cpp]: They are always separate. [Gtest initiates the structure anew between every test](https://github.com/google/googletest/blob/master/googletest/docs/faq.md#should-i-use-the-constructordestructor-of-the-test-fixture-or-setupteardown), which is why we do this everywhere. For example, your own `omp/test/solver/lower_trs_kernels.cpp` use this model.

Comment (pratikvn) [reply cuda/test/solver/lower_trs_kernels.cpp]: But there it uses the `initialize_data()`. I could do the same here.

Comment (tcojean) [reply cuda/test/solver/lower_trs_kernels.cpp]: The `initialize_data()` is here only when you need to control the size of the matrices. When you don't need to, everything could just go into the constructor or directly in the class definition, as, as the link from gtest says, they anyway always create a new test class for every test. The only time we would share anything would only be if we had any static member functions, I suppose.

File: include/ginkgo/core/solver/lower_trs.hpp [hunk 21]
```diff
@@ -196,23 +91,23 @@ class LowerTrs : public EnableLinOp<LowerTrs<ValueType, IndexType>>,
...
      *
-     * @return the preconditioner operator used by the solver
+     * @return the trs solve struct
      */
-    std::shared_ptr<const LinOp> get_preconditioner() const override
+    gko::solver::SolveStruct *get_solve_struct() const
     {
-        return preconditioner_;
+        return solve_struct_.get();
     }
```
Comment (tcojean) [inline include/ginkgo/core/solver/lower_trs.hpp]: You need to test these two (`get_preconditioner` and `get_solve_struct`), although they are trivial.
See [coverity's report](https://codecov.io/gh/ginkgo-project/ginkgo/pull/336/src/include/ginkgo/core/solver/lower_trs.hpp).

You technically even need to test that when you do pass in a preconditioner the preconditioner is actually generated (you get a a non identity I guess?).

File: reference/test/solver/lower_trs_kernels.cpp [hunk 22]
```diff
@@ -72,19 +79,31 @@ class LowerTrs : public ::testing::Test {
...
+    std::unique_ptr<gko::solver::LowerTrs<>::Factory> lower_trs_factory_mrhs;
     std::unique_ptr<gko::solver::LowerTrs<>::Factory> lower_trs_factory_big;
 };
 
 
+TEST_F(LowerTrs, RefLowerTrsFlagCheckIsCorrect)
+{
+    bool trans_flag = true;
+    bool expected_flag = false;
+    gko::kernels::reference::lower_trs::perform_transpose(ref, trans_flag);
```
Comment (thoasm) [inline reference/test/solver/lower_trs_kernels.cpp]: nit: Newline above to separate AAA.

File: include/ginkgo/core/solver/lower_trs.hpp (hunk 21 above)
Comment (pratikvn) [reply include/ginkgo/core/solver/lower_trs.hpp]: I actually do test the second case [here](https://github.com/ginkgo-project/ginkgo/blob/develop/core/test/solver/lower_trs.cpp#L73). But I should add tests for the getters.

File: include/ginkgo/core/solver/lower_trs.hpp [hunk 23]
```diff
@@ -196,23 +91,23 @@ class LowerTrs : public EnableLinOp<LowerTrs<ValueType, IndexType>>,
...
 
     /**
-     * Returns the preconditioner operator used by the solver.
+     * Get the triangular solve struct
      *
-     * @return the preconditioner operator used by the solver
+     * @return the trs solve struct
      */
-    std::shared_ptr<const LinOp> get_preconditioner() const override
+    gko::solver::SolveStruct *get_solve_struct() const
```
[the remark this answers is not in the document]
Comment (pratikvn) [reply include/ginkgo/core/solver/lower_trs.hpp]: Yes, returning a `shared_ptr` would indeed be better.

I just wanted to add some clarification with regard to the namespaces. I think it makes it a bit easier as otherwise, it would not be so clear.

File: core/solver/lower_trs_kernels.hpp [hunk 24]
```diff
@@ -40,26 +40,42 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
 
 
+#define GKO_DECLARE_LOWER_TRS_CHECK_TRANSPOSABILITY_KERNEL()            \
+    void perform_transpose(std::shared_ptr<const DefaultExecutor> exec, \
+                           bool &transposability)
+
+
+#define GKO_DECLARE_LOWER_TRS_INIT_STRUCT_KERNEL()                \
+    void init_struct(std::shared_ptr<const DefaultExecutor> exec, \
+                     std::shared_ptr<gko::solver::SolveStruct> &solve_struct)
```
[the remark this answers is not in the document]
Comment (pratikvn) [reply core/solver/lower_trs_kernels.hpp]: I think I cannot use a plain ptr here because i need to use the shared_ptr constructor to create a new instance of the SolveStruct so that there is a reference count.

File: reference/test/solver/lower_trs.cpp [hunk 25]
```diff
@@ -84,43 +84,35 @@ TEST_F(LowerTrs, LowerTrsFactoryCreatesCorrectSolver)
...
 TEST_F(LowerTrs, CanBeCopied)
 {
-    auto copy = Solver::build().on(exec)->generate(Mtx::create(exec),
-                                                   Mtx::create(exec));
+    auto copy = Solver::build().on(exec)->generate(Mtx::create(exec));
 
     copy->copy_from(gko::lend(lower_trs_solver));
     auto copy_mtx = copy->get_system_matrix();
     auto d_copy_mtx = Mtx::create(exec);
     copy_mtx->convert_to(gko::lend(d_copy_mtx));
```
[the remark this answers is not in the document]
Comment (pratikvn) [reply reference/test/solver/lower_trs.cpp]: I actually dont need this file anymore, so I have removed it.

File: core/solver/lower_trs_kernels.hpp (hunk 24 above)
[the remark this answers is not in the document]
Comment (thoasm) [reply core/solver/lower_trs_kernels.hpp]: That makes sense,

File: cuda/test/solver/lower_trs_kernels.cpp (hunk 14 above)
Comment (pratikvn) [reply cuda/test/solver/lower_trs_kernels.cpp]: Resolved via #343

File: cuda/solver/lower_trs_kernels.cu [hunk 26]
```diff
@@ -65,8 +152,60 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
...
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(
+            cusparseSetPointerMode(handle, CUSPARSE_POINTER_MODE_HOST));
+        if (b->get_stride() == 1) {
+            auto temp_b = const_cast<ValueType *>(b->get_const_values());
+            cusparse::csrsm_solve(
+                handle, CUSPARSE_OPERATION_NON_TRANSPOSE, matrix->get_size()[0],
+                b->get_stride(), &one, solve_struct->factor_descr,
+                matrix->get_const_values(), matrix->get_const_row_ptrs(),
+                matrix->get_const_col_idxs(), solve_struct->solve_info, temp_b,
+                b->get_size()[0], x->get_values(), x->get_size()[0]);
```
[the remark this answers is not in the document]
Comment (pratikvn) [reply cuda/solver/lower_trs_kernels.cu]: So apparently, it works nonetheless. the cusparse analysis function does not seem to use the num_rhs parameter. I will leave that as a parameter because as Terry suggests in the future it is possible that they use this to do this to do something more sophisticated.

So, in short. you can use not set the parameter(default would be 1) and it pass in your multiple rhs vector and it would solve without any problems.

Regarding the fact that it might cause some confusion, I will add an explanation to the parameter doc.

File: include/ginkgo/core/solver/lower_trs.hpp (hunk 23 above)
[the remark this answers is not in the document]
Comment (pratikvn) [reply include/ginkgo/core/solver/lower_trs.hpp]: Actually, not I just remove this getter, as it does not really serve any purpose.

File: cuda/solver/lower_trs_kernels.cu (hunk 26 above)
[the remark this answers is not in the document]
Comment (thoasm) [reply cuda/solver/lower_trs_kernels.cu]: Thanks for testing!
Interesting that it works without any modification. And yes, in that case, documenting this case would be good.

File: include/ginkgo/core/solver/lower_trs.hpp [hunk 27]
```diff
@@ -121,8 +111,11 @@ class LowerTrs : public EnableLinOp<LowerTrs<ValueType, IndexType>>,
         /**
          * Number of right hand sides.
          *
-         * @note: This value must be same as to be passed to the b vector of
-         * apply.
+         * @note: This value is currently a dummy value which is not used by the
+         * analysis step. It is possible that future algorithms (cusparse
+         * csrsm2) make use of the number of right hand sides for a more
+         * sophisticated implementation. Hence this parameter is left here. But
+         * currently, there is no need to use it.
```
Comment (thoasm) [inline include/ginkgo/core/solver/lower_trs.hpp]: Could you make this align with the `@note`? Also, I don't think the colon is necessary (I also don't know if it causes problems either).

File: omp/solver/lower_trs_kernels.cpp [hunk 28]
```diff
@@ -92,7 +92,7 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
 template <typename ValueType, typename IndexType>
 void solve(std::shared_ptr<const OmpExecutor> exec,
            const matrix::Csr<ValueType, IndexType> *matrix,
-           gko::solver::SolveStruct *solve_struct,
+           const gko::solver::SolveStruct *solve_struct,
```
Comment (thoasm) [inline omp/solver/lower_trs_kernels.cpp]: please remove the `gko::`

File: reference/solver/lower_trs_kernels.cpp [hunk 29]
```diff
@@ -88,7 +88,7 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
 template <typename ValueType, typename IndexType>
 void solve(std::shared_ptr<const ReferenceExecutor> exec,
            const matrix::Csr<ValueType, IndexType> *matrix,
-           gko::solver::SolveStruct *solve_struct,
+           const gko::solver::SolveStruct *solve_struct,
```
Comment (thoasm) [inline reference/solver/lower_trs_kernels.cpp]: `gko::`

File: reference/test/solver/lower_trs.cpp [hunk 30]
```diff
@@ -34,12 +34,12 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 
 #include <memory>
+#include <typeinfo>
```
Comment (thoasm) [inline reference/test/solver/lower_trs.cpp]: Why did you add this header? I don't see you using it here.

File: cuda/test/solver/lower_trs_kernels.cpp (hunk 14 above)
Comment (yhmtsai) [reply cuda/test/solver/lower_trs_kernels.cpp]: Both of `cuda` and `gtest` are system third party librarires so you need to put them into same group

File: cuda/solver/lower_trs_kernels.cu [hunk 31]
```diff
@@ -53,10 +66,96 @@ namespace cuda {
...
+            CUSPARSE_OPERATION_TRANSPOSE, matrix->get_size()[0], num_rhs,
+            matrix->get_num_stored_elements(), &one, solve_struct->factor_descr,
+            matrix->get_const_values(), matrix->get_const_row_ptrs(),
+            matrix->get_const_col_idxs(), nullptr, num_rhs,
+            solve_struct->solve_info, solve_struct->policy,
+            &solve_struct->factor_work_size);
+
+        // allocate workspace
+        solve_struct->factor_work_vec =
+            exec->alloc<void *>(solve_struct->factor_work_size);
```
Comment (yhmtsai) [inline cuda/solver/lower_trs_kernels.cu]: Is it possible that allocate the memory before free it?

File: cuda/base/cusparse_bindings.hpp [hunk 32]
```diff
@@ -518,53 +573,13 @@ inline void destroy(cusparseMatDescr_t descr)
 
 inline gko::solver::SolveStruct *init_trs_solve_struct()
```
[the remark this answers is not in the document]
Comment (pratikvn) [reply cuda/base/cusparse_bindings.hpp]: You are right!

File: cuda/solver/lower_trs_kernels.cu (hunk 31 above)
Comment (pratikvn) [reply cuda/solver/lower_trs_kernels.cu]: Do you mean an adding and additional check like this ? 

```c++
        if (solve_struct->factor_work_vec != nullptr) {
            GKO_ASSERT_NO_CUDA_ERRORS(cudaFree(solve_struct->factor_work_vec));
        }

```

File: cuda/solver/lower_trs_kernels.cu [hunk 33]
```diff
@@ -53,10 +66,98 @@ namespace cuda {
...
+
+#endif
+}
+
+
+void init_struct(std::shared_ptr<const CudaExecutor> exec,
+                 std::shared_ptr<solver::SolveStruct> &solve_struct)
+{
+    solve_struct =
+        std::shared_ptr<gko::solver::SolveStruct>(new solver::SolveStruct());
```
[the remark this answers is not in the document]
Comment (pratikvn) [reply cuda/solver/lower_trs_kernels.cu]: Sorry, missed this.

File: include/ginkgo/core/solver/lower_trs.hpp [hunk 34]
```diff
@@ -168,6 +63,11 @@ public:                                                                        \
  * format. If the matrix is not in CSR, then the generate step converts it into
  * a CSR matrix. The generation fails if the matrix is not convertible to CSR.
  *
+ * @note As the constructor uses the copy and convert functionality, it is not
+ * possible to create a empty solver or a solver with a matrix in any other
+ * format other than CSR, if none of the executor modules are being compiled
+ * with.
```
[the remark this answers is not in the document]
Comment (pratikvn) [reply include/ginkgo/core/solver/lower_trs.hpp]: Ah, i misunderstood what you said
